### PR TITLE
refactor: add output type annotations to scalar ops

### PR DESCRIPTION
**Pull request by TrevorBergeron**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

File: bigframes/dtypes.py [hunk 1]
```diff
@@ -42,6 +44,12 @@
     pd.ArrowDtype,
     gpd.array.GeometryDtype,
 ]
+ExpressionType = Union[Dtype, None]
```
Comment (shobsi) [inline bigframes/dtypes.py:47]: [nit] typing.Optional[Dtype] might be more concise

File: bigframes/operations/__init__.py [hunk 2]
```diff
@@ -126,86 +139,101 @@ def _convert_expr_input(
...
         bases=(UnaryOp,),
         frozen=True,
     )()
 
 
-def create_binary_op(name: str) -> BinaryOp:
+def create_binary_op(
+    name: str, type_rule: op_typing.OpTypeRule = op_typing.Supertype()
+) -> BinaryOp:
+    def output_type(self, *input_types: dtypes.ExpressionType) -> dtypes.ExpressionType:
```
Comment (shobsi) [inline bigframes/operations/__init__.py:159]: This seems repetitive all create_*_op methods. Would it make sense to define outside once and reuse? Or even define in a base `class Op` and derive the *naryOp classes from that?

Comment (TrevorBergeron) [reply bigframes/operations/__init__.py]: deduplicated into OpTypeRule class itself. Maybe should jsut have OpTypeRule be a Callable anyways...

File: bigframes/dtypes.py (hunk 1 above)
Comment (TrevorBergeron) [reply bigframes/dtypes.py]: Done

**Review by shobsi — APPROVED** [review]

**Pull request merged**